File: WanderingSalesman/Episode_4/mcherm/javascript/mcherm_episode_4.js
//
// Implementation of Episode 4 of the Wandering Salesman problem
//

const fs = require('fs')

const mapData = JSON.parse(fs.readFileSync("../../../inputdata/small-map.json"));

function countPaths(startNode, length) {
    // --- make a list of counts for each node (for length 1) ---
    let counts = {};
    for (const node of mapData.nodes) {
        counts[node] = BigInt(1);
    }

    // --- find the next level of counts as many times as needed ---
    for (let i = 0; i < length - 1; i++) {
        const newCounts = {}
        for (const node of mapData.nodes) {
            let newCount = BigInt(0);
            for (const neighbor of mapData.neighbors[node]) {
                newCount += counts[neighbor];
            }
            newCounts[node] = newCount;
        }
        counts = newCounts;
    }

    // --- Return the correct answer ---
    return counts[startNode];
}


const pathLen = 300000;
const startTime = performance.now();
const pathCount = countPaths("A", pathLen);
const endTime = performance.now();
console.log(`For length ${pathLen} there are ${pathCount} paths.`);
const seconds = (endTime - startTime) / 1000;
console.log(`That took ${seconds} seconds.`);

File: WanderingSalesman/Episode_4/mcherm/rust/mcherm_episode_4.rs
use std::fs;
use std::collections::HashMap;
use serde::Deserialize;
use serde_json;
use rug::Integer;
use std::time::Instant;


#[derive(Deserialize)]
struct MapDescription {
    nodes: Vec<String>,
    neighbors: HashMap<String,Vec<String>>,
}


fn load_map_description() -> Result<MapDescription, Box<dyn std::error::Error>> {
    let s = fs::read_to_string("inputdata/small-map.json")?;
    let answer = serde_json::from_str(&s)?;
    return Ok(answer);
}



type Node<'a> = &'a str;


/// Given a map and a start position and number of steps, this returns the count
/// of paths of that length starting from that location.
fn count_paths<'a>(map: &'a MapDescription, start_position: Node<'a>, num_steps: u32) -> Integer {
    // --- Start with steps_from for paths of length 1 ---
    let mut steps_from: HashMap<Node<'a>,Integer> = HashMap::new();
    let mut path_len: u32 = 1; // The length that steps_from is valid for
    for node in map.nodes.iter() {
        steps_from.insert(node, Integer::from(1));
    }

    // --- Keep stepping forward steps_from to the next path_len until reaching num_steps ---
    while path_len < num_steps {
        let mut next_steps_from: HashMap<Node<'a>,Integer> = HashMap::new();
        for node in map.nodes.iter() {
            let mut steps_from_this_node = Integer::ZERO;
            let neighbors = map.neighbors.get(node).unwrap();
            for neighbor in neighbors {
                steps_from_this_node += steps_from.get(neighbor.as_str()).unwrap();
            }
            next_steps_from.insert(node, steps_from_this_node);
        }
        steps_from = next_steps_from;
        path_len += 1;
    }

    // --- Return the result ---
    steps_from.get(start_position).unwrap().clone()
}



fn main() -> Result<(), Box<dyn std::error::Error>> {
    let map = load_map_description()?;

    let path_len = 300000;
    let start_time = Instant::now();
    let path_count = count_paths(&map, &"A", path_len);
    let execute_seconds = start_time.elapsed().as_secs();
    println!("For length {} there are {} paths.", path_len, path_count);
    println!("Taking {} seconds.", execute_seconds);
    return Ok(());
}
